### PR TITLE
feat(admin,admin-ui,medusa): Add Medusa Admin plugin

### PR DESCRIPTION
**Pull request by kasperkristensen**
**What**

**`@medusajs/medusa`**
- Adds a new step to the plugin dependency injection lifecycle - `setup`.
  - Adding code inside a directory named `/setup` in your plugin allows you to run function before any loaders are run and API endpoints are injected. For the admin plugin this is necessary to build the admin project before we inject the API endpoint, otherwise, we would be exposing an outdated admin build. This can also have usages in other plugins, and an example would be the need to fetch an external config before running loaders, etc.

**`@medusajs/admin-ui`**
- Migrates the latest develop branch from https://github.com/medusajs/admin to a core package. 
- Exposes a slim API to programmatically rebuild the project.

**`@medusajs/admin`**
- Plugin responsible for exposing the admin dashboard on the server.
- Accepts the following options:
  - `serve?: boolean`: Decides whether to statically serve the dashboard on the server. The default is `true`.
  - `outDir?: string`: If provided the admin UI will be built to the path specified relative to the current working directory. The default is `undefined`.
  - `path?: string`: Allows the user to choose which path they want to expose the dashboard on. Cannot be one of the reserved paths: `["admin", "store"]`, or be suffixed or prefixed with a `/`. The default is `"app"`
  - `backend?: string`: If the user wishes to build the UI with the intent of hosting it separately from the server they should provide the URL for their server here. During the build, the URL will be injected and used for initializing `medusa-react`, and the `axios` client used for `medusaRequest`. The default is `undefined`.

**How to test locally**
The easiest way to test the plugin is by using `yarn link`, this requires that you upgrade your version of yarn in your local Medusa server repository if you haven't already.

```bash
cd my-medusa-server
yarn set version berry
yarn
```

After upgrading and re-installing your dependencies, you can create the necessary links:

```bash
yarn link ../path/to/local/core/repo/packages/admin
yarn link ../path/to/local/core/repo/packages/admin-ui
```

__Note__ that you may also need to link your local versions of `@medusajs/medusa`, `medusa-fulfillment-manual`, and  `medusa-payment-manual`.

In your `medusa-config.js` we can now add the options for the admin plugin:

```js
// This is the place to include plugins. See API documentation for a thorough guide on plugins.
const plugins = [
  `medusa-fulfillment-manual`,
  `medusa-payment-manual`,
  {
    resolve: `medusa-payment-stripe`,
    options: {
      api_key: STRIPE_API_KEY,
      webhook_secret: STRIPE_WEBHOOK_SECRET,
    },
  },
  {
    resolve: "@medusajs/admin",
    /** @type {import('@medusajs/admin').PluginOptions} */
    options: {
      path: "app",
    },
  },
];
```
The type import is optional but recommended for inline documentation on the different options.

You are ready to test the plugin. The first time you run `yarn start` the plugin will automatically build the admin project and expose it on the path that you defined in the options. Any changes to the options will be picked up and trigger a rebuild on subsequent start-ups. If you want to trigger a rebuild yourself you can use the CLI command `medusa-admin build` or add it as a script to your package.json.

```json
"build:admin": "medusa-admin build",
```

**Note**
By agreement with Oli, I have disabled unit tests in `medusa-react` and `admin-ui` for the time being. The tests currently fail in our CI, due to an issue with multiple versions of `react-dom` not being resolved correctly when tests are run in CI. The fix to this is to update the version of `react` and `react-dom` used for testing in `medusa-react`. This will however require updating every test in the project, as we will need to update a test dependency, that has a slightly different API then what we currently use. To not increase the scope of this PR, the tests have therefore been disabled. I'll follow up with a separate PR to fix the tests.

Resolves CORE-757

### CODE REVIEW

**Comment by kasperkristensen** [comment]
@olivermrbl don't think that is related to the admin plugin. Looks to be an issue with this line of code: https://github.com/medusajs/medusa/blob/84b6fe1d9305670bff6b8f6503feeef0d360fcb4/packages/medusa-cli/src/index.js#L6 where we import `semver` in `medusa-cli`, but it's not part of the dependencies of the package (we should fix this but not related to this PR). Did you perhaps also yarn link `medusa-cli`, since the error is thrown from a Yarn Plug'n'Play file?

File: packages/admin/src/commands/run.ts [hunk 1]
```diff
@@ -0,0 +1,8 @@
+import { createCli } from "./create-cli"
```
Comment (olivermrbl) [inline packages/admin/src/commands/run.ts:1]: suggestion(non-blocking): Rename file to `index.ts` to be consistent with our other CLIs. 

Current file name could also be misleading in the sense that some think this exposes a command `run`

**Review by olivermrbl — APPROVED** [review]
This is outstanding work @kasperkristensen 🎆  

I played around with it locally, and it works like a charm. 

Would like to have a couple more reviewers (read environments) test it out 💪

**Comment by olivermrbl** [comment]
Also, please include a changeset 👍

File: packages/medusa/src/loaders/plugins.ts [hunk 2]
```diff
@@ -544,6 +550,28 @@ function registerModels(
...
 }
 
+/**
+ * Runs all setup functions in a plugin. Setup functions are run before anything from the plugin is
+ * registered to the container. This is useful for running custom build logic, fetching remote
+ * configurations, etc.
+ * @param pluginDetails The plugin details including plugin options, version, id, resolved path, etc.
+ */
+async function runSetupFunctions(pluginDetails: PluginDetails): Promise<void> {
+  const files = glob.sync(`${pluginDetails.resolve}/setup/*.js`, {})
```
Comment (riqwan) [inline packages/medusa/src/loaders/plugins.ts:560]: suggestion: should this be picking up `ts` files as well?

**Review by riqwan — APPROVED** [review]
🔥 work! Looks amazing!

File: packages/medusa/src/loaders/plugins.ts (hunk 2 above)
Comment (kasperkristensen) [reply packages/medusa/src/loaders/plugins.ts:560]: We are proccesing the built files so they will always be `*.js`

**Pull request merged**